File: common/src/main/java/cloudcmd/common/engine/LocalBlockCache.java
package cloudcmd.common.engine;

import cloudcmd.common.adapters.Adapter;
import cloudcmd.common.adapters.FileAdapter;
import cloudcmd.common.config.ConfigStorageService;
import org.json.JSONException;

import java.io.File;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.*;

public class LocalBlockCache implements BlockCache
{
  private Adapter _cacheAdapter = null;
  private Map<String, List<Adapter>> _hashProviders = null;

  public void init() throws Exception
  {
    URI adapterUri = new URI("file:///" + ConfigStorageService.instance().getConfigRoot() + File.separator + "cache");
    _cacheAdapter = new FileAdapter();
    _cacheAdapter.init(0, FileAdapter.class.getName(), new HashSet<String>(), adapterUri);
//    loadCache(Integer.MAX_VALUE);
  }

  @Override
  public void shutdown()
  {
  }

  @Override
  public Adapter getBlockCache()
  {
    return _cacheAdapter;
  }

  @Override
  public void refreshCache(int maxTier) throws Exception
  {
    List<Adapter> adapters = new ArrayList<Adapter>(ConfigStorageService.instance().getAdapters());
    adapters.add(_cacheAdapter);

    for (final Adapter adapter : adapters)
    {
      try
      {
        adapter.refreshCache();
      }
      catch (Exception e)
      {
        e.printStackTrace();
      }
    }

    loadCache(maxTier);
  }

  @Override
  public void loadCache(int maxTier) throws Exception
  {
    List<Adapter> adapters = new ArrayList<Adapter>(ConfigStorageService.instance().getAdapters());
    adapters.add(_cacheAdapter);

    _hashProviders = new HashMap<String, List<Adapter>>();

    for (final Adapter adapter : adapters)
    {
      if (adapter.Tier > maxTier) continue;

      Set<String> adapterDescription = adapter.describe();

      for (final String hash : adapterDescription)
      {
        if (!_hashProviders.containsKey(hash))
        {
          _hashProviders.put(hash, new ArrayList<Adapter>());
        }
        _hashProviders.get(hash).add(adapter);
      }
    }
  }

  @Override
  public Map<String, List<Adapter>> getHashProviders()
  {
    if (_hashProviders == null)
    {
      throw new RuntimeException("call loadCache first");
    }
    return Collections.unmodifiableMap(_hashProviders);
  }
}

File: common/src/main/scala/cloudcmd/common/engine/LocalBlockCache.scala
package cloudcmd.common.engine

import cloudcmd.common.adapters.Adapter
import cloudcmd.common.config.ConfigStorage
import java.util

class LocalBlockCache(configStorage : ConfigStorage) extends BlockCache {
  private var _hashProviders: java.util.Map[String, java.util.List[Adapter]] = null

  private def available(p: Adapter, minTier: Int, maxTier: Int) = (p.IsOnLine && p.Tier >= minTier && p.Tier <= maxTier)

  def init {
  }

  def shutdown {}

  def refreshCache(minTier: Int, maxTier: Int) {
    import scala.collection.JavaConversions._
    val adapters : List[Adapter] = configStorage.getAdapters.toList
    adapters.filter(available(_, minTier, maxTier)).par.foreach(_.refreshCache())
  }

  def loadCache(minTier: Int, maxTier: Int) : java.util.Map[String, java.util.List[Adapter]] = {
    _hashProviders = buildHashProviders(minTier, maxTier)
    getHashProviders
  }

  private def buildHashProviders(minTier: Int, maxTier: Int) : java.util.Map[String, java.util.List[Adapter]] = {
    import scala.collection.JavaConversions._

    val adapters : List[Adapter] = configStorage.getAdapters.toList

    val filteredAdapters = adapters.filter(available(_, minTier, maxTier))
    val am = filteredAdapters.par.flatMap(p => p.describe).toSet

    val hashProviders = new java.util.HashMap[String, java.util.List[Adapter]]

    // TODO: this code makes me cry
    am.par.foreach{ hash =>
      val res = new util.ArrayList[Adapter]
      filteredAdapters.foreach(adapter => if (adapter.describe().contains(hash)) res.add(adapter))
      hashProviders.synchronized {
        hashProviders.put(hash, res)
      }
    }

    hashProviders
  }

  def getHashProviders: java.util.Map[String, java.util.List[Adapter]] = {
    if (_hashProviders == null) {
      throw new RuntimeException("call loadCache first")
    }
    java.util.Collections.unmodifiableMap(_hashProviders)
  }
}